docs/perf-profiles.md

```diff
@@ -0,0 +1,109 @@
+
+Azure Disk CSI driver v2 allows customers to tweak the guest OS device settings using perfProfile feature.
+Users can chose from a list of `perfProfile` to tweak the IO behavior of their block device in accordance with their workloads.
+
+`perfProfile` can be set at the `StorageClass` level and applies to all the disks created using the `StorageClass`
+
+## Perf Profiles
+
+Today user can chose from `Basic` and `Manual` `perfProfile`
+
+### Basic
+
+`Basic` `perfProfile` offers a hands free option to tune the device settings for a balanced throughput and TPS workloads.
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: sc-kubestone-perf-optimized-premium-ssd-csi
+provisioner: disk.csi.azure.com
+parameters:
+  skuName: Premium_LRS
+  perfProfile: Basic # available values: None(by default), Basic, Manual. These are case insensitive.
+reclaimPolicy: Delete
+volumeBindingMode: Immediate
+allowVolumeExpansion: true
+```
+
+### Manual
+
+`Manual` `perfProfile` gives users the ultimate flexibility to tweak any device setting they desire, to optimize the disk IOs for their workload.
+
+Setting overrides can be provided in the `parameters` section of the `StorageClass`, using a prefix `disk.csi.azure.com/device-setting/`.
+
+If `perfProfile` is set to `Manual`, driver treats any parameter which starts with prefix `disk.csi.azure.com/device-setting/` as a device setting override.
+
+Device setting overides can be provided in format `<deviceSettingPrefix><deviceSetting>: "<deviceSettingValue>"`.
+
+For example:
+
+To set /dev/sda/queue/scheduler (/dev/sda is an example, a block device root, provisioned for the pod) to mq-deadline, users should use below override in parameter.
+
+```yaml
+  disk.csi.azure.com/device-setting/queue/scheduler: "mq-deadline"
+```
+
+There are no limitations to how many settings user can tweak using this `perfProfile`.
+
+If `perfProfile` is set to `Manual`, atleast one device setting override should be provided in the `parameters`.
+
+Incase, no device setting override is provided and the `perfProfile` is set to `Manual`, CSI driver will fail any disk created using this `StorageClass`.
+
+Any incorrect `deviceSetting` overrides provided in the `parameters` section will result in failure of disk staging on the node and will result in pod scheduling failure.
+
+Users should be careful about using manual `perfProfile` and make sure they test all the `deviceSetting` overrides in linux kernel they plan to use.
+
+Here's an example of a `StorageClass` which tweaks some well known block device settings on linux kernel.
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: sc-kubestone-perf-optimized-premium-ssd-csi
+provisioner: disk.csi.azure.com
+parameters:
+  skuName: Premium_LRS
+  perfProfile: Manual # available values: None(by default), Basic, Manual. These are case insensitive.
+  disk.csi.azure.com/device-setting/queue/max_sectors_kb: "52"
```

Comment:
`StorageClass` parameters are already provisioner-specific, so the `disk.csi.azure.com` prefix seems unnecessary here.

Also, I wonder if it would cleaner/easier to configure using a single value with a YAML encoded string, e.g.:

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: sc-kubestone-perf-optimized-premium-ssd-csi
provisioner: disk.csi.azure.com
parameters:
  perfProfile: Manual
  perfProfileSettings: |
    iosched:
      fifo_batch: true
      writes_starved: true
    queue:
      depth: 16
      max_sectors_kb: 52
      scheduler: "mq-deadline"
      nr_requests: 16
      read_ahead_kb: 8
      wbt_lat_usec: 0
      rotational: false
      nomerges: false
reclaimPolicy: Delete
volumeBindingMode: Immediate
allowVolumeExpansion: true
```

Reply:
Oh, I see: the parameter names after the common prefix are file names under `/sys/block/<device>`. While this allows for ultimate flexibility, this seems kind of dangerous to me. I guess you need to be an admin to create a `StorageClass`, so maybe this is OK. I need to think this through.

Reply:
As discuss offline, given these parameters will make it to PV. I'll keep what we have right now for simplicity.